src/walletSdk/Horizon/Transaction/TransactionBuilder.ts

```diff
@@ -43,8 +41,17 @@ export class TransactionBuilder {
     if (!timebounds) {
       this.builder.setTimeout(cfg.stellar.defaultTimeout);
     }
+  }
 
-    this.sourceAccount = sourceAccount.accountId();
+  sponsoring(
+    sponsorAccount: AccountKeypair,
+    sponsoredAccount?: AccountKeypair,
+  ): SponsoringBuilder {
+    return new SponsoringBuilder(
```

Comment:
note: slight behind the scenes difference from Kotlin sdk [here](https://github.com/stellar/kotlin-wallet-sdk/blob/ac2a0f1d3944d190a287a22b0c7c9d23a666ee9d/wallet-sdk/src/main/kotlin/org/stellar/walletsdk/horizon/transaction/TransactionBuilder.kt#L65-L74).  In Kotlin (if I'm understanding it right) calling .sponsoring returns the same TransactionBuilder instance, after the building function modified the inner operations array.

here we're returning a new SponsoringBuilder instance. This was so instead of the api looking like this:
`txBuilder.sponsoring(sponsorKp, buildingFunction).build()`
(following the lambda function pattern Kotlin is using)

we can do instead:
`txBuilder.sponsoring(sponsorKp).createAccount().build()`

which follows the way the builder looks normally w/o sponsoring

Reply:
How to get back to non-sponsoring builder? 
In kotlin, you can add non-sponsored transactions after 
```kotlin
 stellar
        .transaction(sponsorKeyPair)
        .sponsoring(sponsorKeyPair, newKeyPair) {
          createAccount(newKeyPair)
        }
        .transfer(someOtherAccount) // <- will do a transfer from sponsorKeyPair in the same transaction
```

Reply:
hmm I was under the assumption it would be only a sponsoring transaction or not. How much of a need is it to be able to support allowing both at the same time? 🤔 

A simple solution could be requiring to put non-sponsoring first:
```
 stellar
        .transaction(sponsorKeyPair)
        .transfer(someOtherAccount)
        .sponsoring(sponsorKeyPair, newKeyPair)
        .createAccount()
```


another way, but it would make the code pretty ugly I think, is to do something like:
```
 stellar
        .transaction(sponsorKeyPair)
        .sponsoring(sponsorKeyPair, newKeyPair)
        .createAccount()
        .stopSponsoring()
        .transfer()
```
(would be creating a TransactionBuilder instance, then a Sponsoring one, then a TransactionBuilder one again, unless there's a  cleaner way of doing I'm not thinking of)


I'm in favor of the first option

Reply:
First option makes things like this to be impossible 
```kotlin
    val newKeyPair = wallet.stellar().account().createKeyPair()

    val replaceWith = account.createKeyPair()

    val modifyAccountTransaction =
      stellar
        .transaction(sponsoringKey)
        .sponsoring(sponsoringKey, newKeyPair) {
          createAccount(newKeyPair)
          addAccountSigner(replaceWith, 1)
          lockAccountMasterKey()
        }
        .transfer(newKeyPair.address, XLM, "10")
        .build()
        .sign(newKeyPair)
        .sign(sponsoringKey)

    wallet.stellar().submitTransaction(modifyAccountTransaction)
```
Here's [a transaction](https://stellar.expert/explorer/testnet/tx/a325268ff0ae1c220ee876ad5001d8d90206d32250352be91d3c685b62336710) 
Wallet may sponsor creation and modification, and also fund the new account in one transaction with this approach

Reply:
Can't we accept lambda that takes SponsoringBuilder?
In ts:
```typescript
.sponsoring(sponsoringKey, newKeyPair, spBuilder => {
          spBuilder.createAccount(newKeyPair)
          spBuilder.addAccountSigner(replaceWith, 1)
          spBuilder.lockAccountMasterKey()
        })
```
And internally create SponsoringBuilder, and apply changes from lambda?

Reply:
> First option makes things like this to be impossible

hmm ok, in that case I'd vote to do this way:

```
 stellar
        .transaction(sponsorKeyPair)
        .sponsoring(sponsorKeyPair, newKeyPair)
        .createAccount()
        .stopSponsoring()
        .transfer()
```

> Can't we accept lambda that takes SponsoringBuilder?

We can definitely do it that way, but to me it feels counter intuitive

eg. I want to create an account:
`stellar.transaction(sponsorKeyPair).createAccount()`

actually I want to sponsor it:
`stellar.transaction(sponsorKeyPair).sponsoring().createAccount()`

if we do the lambda function way that code will error. I actually implemented it at first using the lambda function way, and made that mistake a couple times so switched it



Maybe it is more intuitive than I'm thinking though. Lmk what you think, I'm happy to do it either way :)

Reply:
Either option is fine to me, how about we create code samples of 2 alternative options and ask Cassio and Piyal on their opinion how it would be easier to use?

Reply:
https://stellarfoundation.slack.com/archives/C03347FNAHK/p1693338265419979

Reply:
sgtm for using the lambda function way 👍 
changed here: https://github.com/stellar/typescript-wallet-sdk/pull/45/commits/0a28f7e2c9da40fb7237c3c698ec995f91a06520